billing/postbilling.go

```diff
@@ -0,0 +1,325 @@
+package billing
+
+import (
+	"bufio"
+	"encoding/json"
+	. "github.com/journeymidnight/yig-billing/helper"
+	"github.com/journeymidnight/yig-billing/prometheus"
+	"github.com/journeymidnight/yig-billing/redis"
+	"github.com/journeymidnight/yig-billing/spark"
+
+	"io/ioutil"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	BillingTypeAPI          = "API"
+	BillingTypeTraffic      = "TRAFFIC"
+	BillingTypeDataRetrieve = "DATARETRIEVE"
+	LoggerBucketInfo        = "PLUGIN LOG"
+	LoggerBuckteError       = "WARRING BUT NOT IMPORTANT"
+)
+
+type BatchUsage struct {
+	RegionId  string      `json:"regionId"`
+	BeginTime string      `json:"beginTime"`
+	EndTime   string      `json:"endTime"`
+	Data      []UserUsage `json:"data"`
+}
+
+type UserUsage struct {
+	ProjectId string         `json:"projectId"`
+	Usages    []BillingUsage `json:"usages"`
+}
+
+type BillingUsage struct {
+	BillType string `json:"billType"`
+	Usage    uint64 `json:"usage"`
+}
+
+const folderLayoutStr = "2006010215"
+
+func postBilling() {
+	wg := new(sync.WaitGroup)
+	Logger.Println("[INFO] Begin to runBilling", time.Now().Format("2006-01-02 15:04:05"))
+	task := new(Task)
+	task.PidCache = make(map[string][]BillingUsage)
+	task.RedisUsageCache = make(map[string]BillingUsage)
+
+	// Get Usages by '.csv' file exported by TiSpark
+	task.ConstructUsageData(Conf.TisparkShell)
+	// Get Traffic From Prometheus
+	task.ConstructTrafficData()
+	// Get Standard_IA DataRetrieve from Prometheus
+	task.ConstructRetrieveStandardIaData()
+	// Get API Count From Prometheus
+	task.ConstructAPIData()
+	// If Enable Usage Cache Is On, Cache Usage to Redis
+	if Conf.EnableUsageCache {
+		wg.Add(1)
+		go task.cacheUsageToRedis(wg)
+	}
+
+	// Constuct BillingData
+	task.BillingData.RegionId = Conf.RegionId
+	hd, _ := time.ParseDuration("-1h")
+	task.BillingData.BeginTime = time.Now().Add(hd).Format(spark.TimeLayoutStr) + ":00:00"
+	task.BillingData.EndTime = time.Now().Format(spark.TimeLayoutStr) + ":00:00"
```

Comment:
Format with `Format`, I don't think the string concatenation is necessary.

Reply:
This is used to mark the time period